src/main/java/me/botsko/prism/listeners/PrismBlockEvents.java

```diff
@@ -27,508 +43,507 @@
-		if (from.getType() == Material.LAVA) {
-			if (Prism.getIgnore().event("lava-flow", event.getBlock())) {
-				RecordingQueue.addToQueue(ActionFactory.createBlock("lava-flow", event.getBlock(), "Lava"));
-			}
-		}
-	}
+    private final Prism plugin;
+
+    /**
+     * Constructor.
+     * @param plugin Prism.
+     */
+    public PrismBlockEvents(Prism plugin) {
+        this.plugin = plugin;
+    }
+
+    /**
+     * Process over each item.
+     * @param block Block
+     * @param callback callback that consumes items/integer pairs.
+     */
+    void forEachItem(Block block, BiConsumer<ItemStack, Integer> callback) {
+        if (block.getType().equals(Material.JUKEBOX)) {
+            final Jukebox jukebox = (Jukebox) block.getState();
+            final Material playing = jukebox.getPlaying();
+            if (playing.equals(Material.AIR)) {
+                return;
+            }
+            final ItemStack i = new ItemStack(jukebox.getPlaying(), 1);
+            callback.accept(i, 0);
+            return;
+        }
+        if (block.getState() instanceof InventoryHolder) {
+            InventoryHolder container = (InventoryHolder) block.getState();
+
+            if (container instanceof DoubleChest) {
+                if (((Chest) block.getBlockData()).getType() == Type.LEFT) {
+                    container = ((DoubleChest) container).getLeftSide();
+                } else {
+                    container = ((DoubleChest) container).getRightSide();
+                }
+            }
+
+            int slot = 0;
+            if (container != null) {
+                for (final ItemStack i : container.getInventory().getContents()) {
+                    // when double chests are broken, they record *all* contents
+                    // even though only half of the chest breaks.
+                    if ((block.getType().equals(Material.CHEST) || block.getType().equals(Material.TRAPPED_CHEST))
+                            && slot > 26)
+                        break;
+                    // record item
+                    if (i != null) {
+                        callback.accept(i, slot);
+                    }
+                    slot++;
+                }
+            }
+        }
+    }
+
+    private void logBlockRelationshipsForBlock(Player player, Block block) {
+        relatedBlockCallback(block, b -> RecordingQueue.addToQueue(
+                ActionFactory.createBlock("block-break", b, player)), s -> plugin.preplannedBlockFalls.put(
+                        s, player.getUniqueId().toString()));
+    }
+
+    void logBlockRelationshipsForBlock(String nonPlayer, Block block) {
+        relatedBlockCallback(block, b -> RecordingQueue.addToQueue(
+                ActionFactory.createBlock("block-break", b, nonPlayer)), s -> plugin.preplannedBlockFalls.put(
+                        s, nonPlayer));
+    }
+
+    private void relatedBlockCallback(Block block, Consumer<Block> breakCallback, Consumer<String> fallCallback) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'relatedBlockCallback(Block, Consumer, Consumer)' has an NPath complexity of 2430](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5438808)